File: src/lib/connection-presets.test.ts
import { describe, expect, it } from 'vitest';
import {
	inferLocation,
	isLocalHost,
	securityLabelForMode,
	transportPresetForLocation
} from '$lib/connection-presets';

describe('connection presets', () => {
	it('maps local databases to disabled transport', () => {
		expect(transportPresetForLocation('local')).toMatchObject({
			mode: 'Disabled',
			legacy_implicit: false
		});
	});

	it('maps cloud and organization databases to verified identity transport', () => {
		expect(transportPresetForLocation('cloud').mode).toBe('VerifyIdentity');
		expect(transportPresetForLocation('organization').mode).toBe('VerifyIdentity');
	});

	it('labels transport security modes consistently', () => {
		expect(securityLabelForMode('VerifyIdentity')).toEqual({
			tone: 'success',
			text: 'Secure connection'
		});
		expect(securityLabelForMode('VerifyCa')).toEqual({
			tone: 'success',
			text: 'Secure connection'
		});
		expect(securityLabelForMode('EncryptOnly')).toEqual({
			tone: 'warning',
			text: 'Encrypted (certificate not verified)'
		});
		expect(securityLabelForMode('Disabled')).toEqual({
			tone: 'danger',
			text: 'Not encrypted'
		});
		expect(securityLabelForMode('Disabled', 'localhost')).toEqual({
			tone: 'danger',
			text: 'Not encrypted - local only'
		});
		expect(securityLabelForMode('Disabled', 'db.example.com')).toEqual({
			tone: 'danger',
			text: 'Not encrypted'
		});
	});

	it('infers only loopback hosts as local', () => {
		expect(isLocalHost('localhost')).toBe(true);
		expect(isLocalHost('127.0.0.1')).toBe(true);
		expect(isLocalHost('[::1]')).toBe(true);
		expect(inferLocation('db.example.com')).toBe('cloud');
	});
});

File: shared/src/test/kotlin/com/safedb/connection/ConnectionPresetsTest.kt
package com.safedb.connection

import com.safedb.model.TransportSecurityMode
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class ConnectionPresetsTest {
    @Test
    fun isLocalHostRecognizesLoopbackForms() {
        assertTrue(isLocalHost("localhost"))
        assertTrue(isLocalHost("127.0.0.1"))
        assertTrue(isLocalHost("::1"))
        assertTrue(isLocalHost("[::1]"))
        assertTrue(isLocalHost("  LOCALHOST  "))
    }

    @Test
    fun isLocalHostRejectsRemoteHosts() {
        assertFalse(isLocalHost("db.example.com"))
        assertFalse(isLocalHost("10.0.0.5"))
    }

    @Test
    fun inferLocationUsesLocalForLoopback() {
        assertEquals(DatabaseLocation.Local, inferLocation("localhost"))
        assertEquals(DatabaseLocation.Cloud, inferLocation("db.example.com"))
    }

    @Test
    fun transportPresetForLocation() {
        assertEquals(
            TransportSecurityMode.Disabled,
            transportPresetForLocation(DatabaseLocation.Local).mode,
        )
        assertEquals(
            TransportSecurityMode.VerifyIdentity,
            transportPresetForLocation(DatabaseLocation.Cloud).mode,
        )
        assertEquals(
            TransportSecurityMode.VerifyIdentity,
            transportPresetForLocation(DatabaseLocation.Organization).mode,
        )
    }
}
